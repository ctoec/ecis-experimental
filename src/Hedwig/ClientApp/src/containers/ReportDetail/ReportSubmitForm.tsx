import React, { useState, FormEvent, useContext, SetStateAction, Dispatch } from 'react';
import Alert, { AlertProps } from '../../components/Alert/Alert';
import { CdcReport, ApiOrganizationsOrgIdReportsIdPutRequest, CdcReportFromJSON } from '../../generated';
import { Mutate } from '../../hooks/useApi';
import UserContext from '../../contexts/User/UserContext';
import idx from 'idx';
import TextInput from '../../components/TextInput/TextInput';
import Checkbox from '../../components/Checklist/Checkbox';
import { AppContext } from '../App/App';
import currencyFormatter from '../../utils/currencyFormatter';
import parseCurrencyFromString from '../../utils/parseCurrencyFromString';
import getIdForUser from '../../utils/getIdForUser';

export type ReportSubmitFormProps = {
  report: CdcReport,
  mutate: Mutate<CdcReport>,
  setAlert: Dispatch<SetStateAction<AlertProps | null>>,
  canSubmit: boolean
};

export default function ReportSubmitForm({ report, mutate, setAlert, canSubmit }: ReportSubmitFormProps) {
  const [accredited, setAccredited] = useState(report.accredited);
  const [c4KRevenue, setC4KRevenue] = useState(report.c4KRevenue || null);
  const [retroactiveC4KRevenue, setRetroactiveC4KRevenue] = useState(report.retroactiveC4KRevenue);
  const [familyFeesRevenue, setFamilyFeesRevenue] = useState(report.familyFeesRevenue || null);

  const { user } = useContext(UserContext);
  const { invalidateCache: invalidateAppCache } = useContext(AppContext);

  const params: ApiOrganizationsOrgIdReportsIdPutRequest = {
    id: report.id || 0,
    orgId: getIdForUser(user, "org")
  };

  function updatedReport(): CdcReport {
    return {
      ...report,
      accredited,
      c4KRevenue: c4KRevenue !== null ? c4KRevenue : undefined,
      retroactiveC4KRevenue,
      familyFeesRevenue: familyFeesRevenue !== null ? familyFeesRevenue : undefined,
    };
  }

  function onSubmit(e: FormEvent) {
    e.preventDefault();
    mutate(
      api => api.apiOrganizationsOrgIdReportsIdPut({
        ...params,
        cdcReport: updatedReport()
      })
    )
      .then(res => {
        if (res) {
          setAlert({
            type: 'success',
            heading: 'Report submitted',
            text: 'You have successfully submitted this report'
          });
          invalidateAppCache(); // Updates the count of unsubmitted reports in the nav bar
        }
      })
      .catch(err => {
        console.log(err);
        setAlert({
          type: 'error',
          heading: 'Report not submitted',
          text: 'There was an error submitting this report'
        })
      })
  }

  return (
    <React.Fragment>
      <form className="usa-form" onSubmit={onSubmit}>
        {report.submittedAt && (
          <p>
            <b>Submitted At:</b> {report.submittedAt.toLocaleDateString()}{' '}
          </p>
        )}
        <div className="usa-checkbox">
          <input
            className="usa-checkbox__input"
            id="accredited"
            type="checkbox"
            disabled={!!report.submittedAt}
            defaultChecked={accredited}
            onChange={e => setAccredited(e.target.checked)}
          />
          <label className="usa-checkbox__label" htmlFor="accredited">
            Accredited
					</label>
        </div>
        <fieldset className="usa-fieldset">
          <legend>
            <h2 className="margin-bottom-0 margin-top-2">Other Revenue</h2>
          </legend>
          <TextInput
            id="c4k-revenue"
            label={
              <React.Fragment>
                <span className="text-bold">Care 4 Kids</span>
              </React.Fragment>
            }
            defaultValue={currencyFormatter(c4KRevenue)}
            onChange={(e) => setC4KRevenue(parseCurrencyFromString(e.target.value))}
            onBlur={event => (event.target.value = c4KRevenue ? currencyFormatter(c4KRevenue) : '')}
            disabled={!!report.submittedAt}
          />
          <div className="margin-top-2">
            <Checkbox
              text="Includes retroactive payment for past months"
              value="c4k-includes-retroactive"
              name="c4k-includes-retroactive"
              onChange={(e) => setRetroactiveC4KRevenue(!!e.target.checked)}
              disabled={!!report.submittedAt}
              checked={retroactiveC4KRevenue}
            />
          </div>
          <TextInput
            id="family-fees-revenue"
            label={<span className="text-bold">Family Fees</span>}
            defaultValue={currencyFormatter(familyFeesRevenue)}
            onChange={(e) => setFamilyFeesRevenue(parseCurrencyFromString(e.target.value))}
            onBlur={event => (event.target.value = familyFeesRevenue ? currencyFormatter(familyFeesRevenue) : '')}
            disabled={!!report.submittedAt}
          />
        </fieldset>
        {!report.submittedAt && <input className="usa-button" type="submit" value="Submit" disabled={!canSubmit} />}
      </form>
    </React.Fragment>
  );
}
