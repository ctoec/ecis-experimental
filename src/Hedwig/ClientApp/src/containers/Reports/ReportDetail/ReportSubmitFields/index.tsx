export * from './Accredited';
export * from './C4KRevenue';
export * from './Comment';
export * from './FamilyFeesRevenue';
export * from './OtherRevenueFieldSet';
export * from './RetroactiveC4KRevenue';
export * from './TimeSplitUtilizationField';
