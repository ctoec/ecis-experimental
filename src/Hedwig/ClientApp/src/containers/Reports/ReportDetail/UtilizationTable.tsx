import React from 'react';
import Table, { TableProps } from '../../../components/Table/Table';
import { CdcReport, Age, FundingTime, Region } from '../../../generated';
import idx from 'idx';
import moment from 'moment';
import { CdcRates } from './CdcRates';
import { prettyAge } from '../../../utils/ageGroupUtils';
import { prettyFundingTime } from '../../../utils/fundingTimeUtils';
import currencyFormatter from '../../../utils/currencyFormatter';

export function calculateRate(accredited: boolean, titleI: boolean, region: Region, ageGroup: Age, time: FundingTime) {
  const rate = CdcRates.find(rate =>
    rate.accredited === accredited &&
    rate.titleI === titleI &&
    rate.region === region &&
    rate.ageGroup === ageGroup &&
    rate.time === time
  );

  return rate ? rate.rate : 0;
}

interface UtilizationTableRow {
  key: string
  ageGroup?: Age
  fundingTime?: FundingTime
  count: number
  capacity: number
  rate?: number
  total: number
  balance: number
}

export default function UtilizationTable(report: CdcReport) {
  const periodStart = idx(report, _ => _.reportingPeriod.periodStart);
  const periodEnd = idx(report, _ => _.reportingPeriod.periodEnd);
  const weeksInPeriod = periodStart && periodEnd ? moment(periodEnd).add(1, 'day').diff(periodStart, 'weeks') : 0;

  const accredited = report.accredited;
  const fundingSpaces = idx(report, _ => _.organization.fundingSpaces) || [];

  const sites = idx(report, _ => _.organization.sites) || [];
  const ageGroups = [Age.InfantToddler, Age.Preschool, Age.SchoolAge];
  const fundingTimes = [FundingTime.Full, FundingTime.Part];

  const rows: UtilizationTableRow[] = sites.flatMap(site => {
    const region = site.region;
    const titleI = site.titleI;

    const enrollments = site.enrollments.flatMap(enrollment => {
      const cdcFunding = enrollment.fundings.find(funding => funding.source === report.type);
      if (!cdcFunding) { return []; }
      return {
        ageGroup: enrollment.ageGroup,
        time: cdcFunding.time,
      };
    });

    return ageGroups.flatMap(ageGroup => fundingTimes.flatMap(fundingTime => {
      const count = enrollments
        .filter(enrollment => enrollment.ageGroup === ageGroup && enrollment.time === fundingTime)
        .length;

      const capacity = idx(
        fundingSpaces.find(space => space.ageGroup === ageGroup && space.time === fundingTime),
        _ => _.capacity
      ) || 0;

      const cappedCount = Math.min(count, capacity);

      const rate = calculateRate(accredited, titleI, region, ageGroup, fundingTime);

      const total = cappedCount * rate * weeksInPeriod;
      const paid = capacity * rate * weeksInPeriod;
      const balance = total - paid;

      return {
        key: `${ageGroup}-${fundingTime}`,
        ageGroup,
        fundingTime,
        count,
        capacity,
        rate,
        total,
        balance,
      }
    }));
  });

  const totalRow = rows.reduce((total, row) => {
    return {
      ...total,
      count: total.count + row.count,
      capacity: total.capacity + row.capacity,
      total: total.total + row.total,
      balance: total.balance + row.balance,
    };
  }, { key: 'total', count: 0, capacity: 0, total: 0, balance: 0 });

  rows.push(totalRow);

  const tableProps: TableProps<UtilizationTableRow> = {
    id: 'utilization-table',
    data: rows.filter(row => row.count || row.capacity),
    rowKey: row => row.key,
    columns: [
      {
        name: '',
        cell: ({ row }) => (
          <th>
            <strong>
              {row.key === 'total' ? 'Total' : `${prettyAge(row.ageGroup)}`}
            </strong>
            {row.fundingTime && (
              <> &ndash; {prettyFundingTime(row.fundingTime)}</>
            )}
          </th>
        )
      },
      {
        name: 'Utilization',
        cell: ({ row }) => (
          <td className={
            (row.key === "total" ? "oec-table__cell--strong" : "") + " " +
            (row.count !== row.capacity ? "oec-table__cell--red" : "")
          }>
            {row.count}/{row.capacity} spaces
          </td>
        )
      },
      {
        name: <>Reimbursement <br />rate (weekly)</>,
        cell: ({ row }) => (
          <td>
            {row.key !== 'total' && currencyFormatter(row.rate || 0)}
          </td>
        )
      },
      {
        name: `Total (${weeksInPeriod} weeks)`,
        cell: ({ row }) => (
          <td className={row.key === "total" ? "oec-table__cell--strong" : ""}>
            {currencyFormatter(row.total)}
          </td>
        )
      },
      {
        name: `Balance`,
        cell: ({ row }) => (
          <td className={
            (row.key === "total" ? "oec-table__cell--strong" : "") + " " +
            (row.balance < 0 ? "oec-table__cell--red" : "")
          }>
            {row.balance < 0 ? '(' : ''}
            {currencyFormatter(Math.abs(row.balance))}
            {row.balance < 0 ? ')' : ''}
          </td>
        )
      }
    ]
  }

  return <>
    <h2 className="margin-bottom-neg-205">Monthly utilization</h2>
    <Table {...tableProps} fullWidth />
  </>
}
