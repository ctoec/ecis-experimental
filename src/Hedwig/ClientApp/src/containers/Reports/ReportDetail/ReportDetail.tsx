import React, { useContext } from 'react';
import { useParams } from 'react-router-dom';
import ReportSubmitForm from './ReportSubmitForm';
import dateFormatter from '../../../utils/dateFormatter';
import UserContext from '../../../contexts/User/UserContext';
import { getIdForUser, reportingPeriodFormatter } from '../../../utils/models';
import useApi from '../../../hooks/useApi';
import { Enrollment } from '../../../generated/models/Enrollment';
import { Button, AlertProps, DirectionalLinkProps, Tag } from '../../../components';
import { DeepNonUndefineable } from '../../../utils/types';
import CommonContainer from '../../CommonContainer';

export default function ReportDetail() {
	const { id } = useParams();
	const { user } = useContext(UserContext);
	const reportParams = {
		id: parseInt(id || '0'),
		orgId: getIdForUser(user, 'org'),
		include: ['organizations', 'enrollments', 'sites', 'funding_spaces', 'child'],
	};
	const [loading, error, report, mutate] = useApi(
		api => api.apiOrganizationsOrgIdReportsIdGet(reportParams),
		[user]
	);

	if (loading || !report) {
		return <div className="Report"></div>;
	}

	const numEnrollmentsMissingInfo = (report.enrollments|| []).filter<DeepNonUndefineable<Enrollment>>(
		e => !!e.validationErrors && e.validationErrors.length > 0
	).length;

	let additionalAlerts: AlertProps[] = [];

	if (error) {
		additionalAlerts.push(
			{
	      type: 'error',
	      heading: 'Something went wrong',
	      text: 'There was an error loading the report',
			},
		);
	}

	if (numEnrollmentsMissingInfo) {
		additionalAlerts.push(
			{
				type: 'error',
				heading: 'Update roster',
				text: `There are ${numEnrollmentsMissingInfo} enrollments missing information required to submit this CDC report.`,
				actionItem: <Button text="Update roster" href="/roster" />,
			},
		);
	}

	const directionalLinkProps: DirectionalLinkProps = {
		direction: 'left',
		to: '/reports',
		text: 'Back to reports',
	};

	return (
		<CommonContainer
			additionalAlerts={additionalAlerts || undefined}
			directionalLinkProps={directionalLinkProps}
		>
			<div className="grid-container">
				<div className="grid-row flex-first-baseline flex-space-between">
					<h1 className="tablet:grid-col-auto">
						{reportingPeriodFormatter(report.reportingPeriod)} {report.type} Report{' '}
						{!report.submittedAt && (
							<Tag text="DRAFT" color="gold-20v" className="margin-left-1 text-middle" />
						)}
					</h1>
					<div className="tablet:grid-col-auto print-btn">
						<Button text="Print" onClick={window.print} appearance="outline" />
					</div>
				</div>
				<p className="usa-intro">
					{report.organization.name} <br />
					{dateFormatter(report.reportingPeriod.periodStart)}â€“
					{dateFormatter(report.reportingPeriod.periodEnd)}
				</p>
				<ReportSubmitForm
					report={report}
					mutate={mutate}
					canSubmit={numEnrollmentsMissingInfo === 0}
				/>
			</div>
		</CommonContainer>
	);
}
