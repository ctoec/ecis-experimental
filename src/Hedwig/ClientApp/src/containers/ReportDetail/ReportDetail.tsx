import React, { useState, useContext } from 'react';
import idx from 'idx';
import { useParams } from 'react-router-dom';
import ReportSubmitForm from './ReportSubmitForm';
import monthFormatter from '../../utils/monthFormatter';
import dateFormatter from '../../utils/dateFormatter';
import DirectionalLink from '../../components/DirectionalLink/DirectionalLink';
import Tag from '../../components/Tag/Tag';
import UserContext from '../../contexts/User/UserContext';
import getIdForUser from '../../utils/getIdForUser';
import useApi from '../../hooks/useApi';
import UtilizationTable from './UtilizationTable';

export default function ReportDetail() {
  let { id } = useParams();
  const { user } = useContext(UserContext);
  const reportParams = {
    id: parseInt(id || '0'),
    orgId: getIdForUser(user, 'org'),
    include: ['organizations', 'enrollments', 'sites', 'funding_spaces'],
  };
  const [loading, error, report] = useApi(
    api => api.apiOrganizationsOrgIdReportsIdGet(reportParams),
    [user]
  );

  console.log(report);

  if (!report) {
    return <div className="Report"></div>;
  }

  return (
    <div className="Report">
      <section className="grid-container">
        <DirectionalLink direction="left" to="/reports" text="Back to reports" />
        <h1>
          {monthFormatter(idx(report, _ => _.reportingPeriod.period))} {report.type} Report{' '}
          {!report.submittedAt && (
            <Tag text="DRAFT" color="gold-20v" addClass="margin-left-1 text-middle" />
          )}
        </h1>
        <p className="usa-intro">
          {idx(report, _ => _.organization.name)} | {dateFormatter(idx(report, _ => _.reportingPeriod.periodStart))} -{' '}
          {dateFormatter(idx(report, _ => _.reportingPeriod.periodEnd))}
        </p>
        <UtilizationTable {...report} />
        <ReportSubmitForm {...report} />
      </section>
    </div>
  );
}
