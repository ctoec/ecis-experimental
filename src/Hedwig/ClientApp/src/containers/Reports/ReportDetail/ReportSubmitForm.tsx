import React, { useState, useEffect, useContext, useReducer } from 'react';
import {
	CdcReport,
	ApiOrganizationsOrgIdReportsIdPutRequest,
	ApiOrganizationsOrgIdEnrollmentsGetRequest,
} from '../../../generated';
import useApi, { Mutate, ApiError } from '../../../hooks/useApi';
import UserContext from '../../../contexts/User/UserContext';
import { Button, TextInput, ChoiceList, FieldSet, ErrorBoundary } from '../../../components';
import AppContext from '../../../contexts/App/AppContext';
import currencyFormatter from '../../../utils/currencyFormatter';
import parseCurrencyFromString from '../../../utils/parseCurrencyFromString';
import { getIdForUser, activeC4kFundingAsOf } from '../../../utils/models';
import UtilizationTable from './UtilizationTable';
import AlertContext from '../../../contexts/Alert/AlertContext';
import { useHistory } from 'react-router';
import { DeepNonUndefineable } from '../../../utils/types';
import {
	useFocusFirstError,
	serverErrorForField,
	isBlockingValidationError,
} from '../../../utils/validations';
import usePromiseExecution from '../../../hooks/usePromiseExecution';
import { reportSubmittedAlert, reportSubmitFailAlert } from '../../../utils/stringFormatters';
import pluralize from 'pluralize';
import { validationErrorAlert } from '../../../utils/stringFormatters/alertTextMakers';
import { FormReducer, formReducer, updateData } from '../../../utils/forms/form';

export type ReportSubmitFormProps = {
	report: DeepNonUndefineable<CdcReport>;
	mutate: Mutate<CdcReport>;
	error: ApiError | null;
	canSubmit: boolean;
};

export default function ReportSubmitForm({
	report,
	mutate,
	error,
	canSubmit,
}: ReportSubmitFormProps) {
	const history = useHistory();

	const { user } = useContext(UserContext);

	const [_report, updateReport] = useReducer<FormReducer<DeepNonUndefineable<CdcReport>>>(
		formReducer,
		report
	);
	const updateFormData = updateData<DeepNonUndefineable<CdcReport>>(updateReport);
	const {
		submittedAt,
		accredited,
		c4KRevenue,
		retroactiveC4KRevenue,
		familyFeesRevenue,
		comment,
	} = _report;

	const { invalidateCache: invalidateAppCache } = useContext(AppContext);
	const { getAlerts, setAlerts } = useContext(AlertContext);
	const alerts = getAlerts();

	const params: ApiOrganizationsOrgIdReportsIdPutRequest = {
		id: report.id || 0,
		orgId: getIdForUser(user, 'org'),
	};

	const enrollmentParams: ApiOrganizationsOrgIdEnrollmentsGetRequest = {
		orgId: getIdForUser(user, 'org'),
		include: ['fundings'],
		startDate: report.reportingPeriod.periodStart,
		endDate: report.reportingPeriod.periodEnd,
		asOf: submittedAt || undefined,
	};

	const [, , allEnrollments] = useApi(
		api => api.apiOrganizationsOrgIdEnrollmentsGet(enrollmentParams),
		[user, report]
	);

	const [care4KidsCount, setCare4KidsCount] = useState(0);

	useEffect(() => {
		if (allEnrollments) {
			var c4kFundedEnrollments = allEnrollments.filter(
				enrollment => !!activeC4kFundingAsOf(enrollment.fundings, submittedAt || undefined)
			);
			var childIds: string[] = [];
			c4kFundedEnrollments.forEach(enrollment => {
				const childId = enrollment.childId;
				if (childIds.indexOf(childId) < 0) {
					childIds.push(childId);
				}
			});
			setCare4KidsCount(childIds.length);
		}
	}, [allEnrollments]);

	// set up form state
	useFocusFirstError([error]);
	const [hasAlertedOnError, setHasAlertedOnError] = useState(false);
	useEffect(() => {
		if (error && !hasAlertedOnError) {
			if (!isBlockingValidationError(error)) {
				throw new Error(error.title || 'Unknown api error');
			}
			setAlerts([validationErrorAlert]);
		}
	}, [error, hasAlertedOnError]);

	function _onSubmit() {
		return mutate(api =>
			api.apiOrganizationsOrgIdReportsIdPut({
				...params,
				cdcReport: { ..._report },
			})
		).then(res => {
			if (res) {
				const newAlert = reportSubmittedAlert(report.reportingPeriod);
				const newAlerts = [...alerts, newAlert];
				setAlerts(newAlerts);
				invalidateAppCache(); // Updates the count of unsubmitted reports in the nav bar
				history.push('/reports', newAlerts);
			}
		});
	}
	const { isExecuting: isMutating, setExecuting: onSubmit } = usePromiseExecution(_onSubmit);

	return (
		<ErrorBoundary alertProps={reportSubmitFailAlert}>
			{report.submittedAt && (
				<p>
					<b>Submitted:</b> {report.submittedAt.toLocaleDateString()}{' '}
				</p>
			)}
			<ChoiceList
				type="check"
				id="accredited"
				name="accredited"
				legend="Accredited"
				disabled={!!report.submittedAt}
				selected={accredited ? ['accredited'] : undefined}
				options={[
					{
						text: 'Accredited',
						value: 'accredited',
					},
				]}
				onChange={updateFormData((_, e) => e.target.checked)}
				className="margin-bottom-5"
			/>
			<UtilizationTable {...{ ..._report, accredited }} />
			<form className="usa-form" onSubmit={onSubmit} noValidate autoComplete="off">
				<h2>Other Revenue</h2>
				<FieldSet id="other-revenue" legend="Other Revenue">
					<TextInput
						type="input"
						id="c4k-revenue"
						name="c4KRevenue"
						label={
							<React.Fragment>
								<span className="text-bold">Care 4 Kids</span>
								<span>
									{' '}
									({care4KidsCount} {pluralize('kid', care4KidsCount)} receiving subsidies)
								</span>
							</React.Fragment>
						}
						defaultValue={currencyFormatter(c4KRevenue)}
						onChange={updateFormData(parseCurrencyFromString)}
						onBlur={event =>
							(event.target.value = c4KRevenue !== null ? currencyFormatter(c4KRevenue) : '')
						}
						optional={true}
						disabled={!!report.submittedAt}
						status={serverErrorForField(
							hasAlertedOnError,
							setHasAlertedOnError,
							'report.c4krevenue',
							error,
							'This information is required for the report'
						)}
					/>
					<ChoiceList
						type="check"
						id="c4k-includes-retroactive"
						legend="Includes retroactive payment"
						name="retroactiveC4KRevenue"
						selected={retroactiveC4KRevenue ? ['retroactiveC4KRevenue'] : undefined}
						onChange={updateFormData((_, e) => e.target.checked)}
						disabled={!!report.submittedAt}
						options={[
							{
								text: 'Includes retroactive payment for past months',
								value: 'retroactiveC4KRevenue',
							},
						]}
					/>
					<TextInput
						type="input"
						id="family-fees-revenue"
						name="familyFeesRevenue"
						label={<span className="text-bold">Family Fees</span>}
						defaultValue={currencyFormatter(familyFeesRevenue)}
						onChange={updateFormData(parseCurrencyFromString)}
						onBlur={event =>
							(event.target.value =
								familyFeesRevenue !== null ? currencyFormatter(familyFeesRevenue) : '')
						}
						disabled={!!report.submittedAt}
						status={serverErrorForField(
							hasAlertedOnError,
							setHasAlertedOnError,
							'familyfeesrevenue',
							error,
							'This information is required'
						)}
					/>
					<TextInput
						type="textarea"
						id="cdc-report-comment"
						name="comment"
						label={
							<span className="text-bold">
								Anything to share with the Office of Early Childhood about your report?
							</span>
						}
						defaultValue={comment || ''}
						onChange={updateFormData()}
						disabled={!!report.submittedAt}
						optional={true}
					/>
				</FieldSet>
				{!report.submittedAt && (
					<Button
						onClick="submit"
						text={isMutating ? 'Submitting...' : 'Submit'}
						disabled={!canSubmit || isMutating}
					/>
				)}
			</form>
		</ErrorBoundary>
	);
}
