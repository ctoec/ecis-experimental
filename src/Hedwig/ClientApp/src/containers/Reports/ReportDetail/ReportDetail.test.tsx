import React from 'react';
import { mount } from 'enzyme';
import 'react-dates/initialize';
import ReportDetail from './ReportDetail';
import CommonContextProviderMock from '../../../contexts/__mocks__/CommonContextProviderMock';

jest.mock('../../../hooks/useApi');
import useApi from '../../../hooks/useApi';

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
	useParams: () => ({
		id: 1,
		orgId: 1,
	}),
}));

afterAll(() => {
	jest.resetModules();
});

describe('ReportDetail', () => {
	it('matches snapshot', () => {
		const wrapper = mount(
			<CommonContextProviderMock>
				<ReportDetail />
			</CommonContextProviderMock>
		);
		expect(wrapper.html()).toMatchSnapshot();
		wrapper.unmount();
	});

	it('shows an alert if enrollments are missing info', () => {
		const wrapper = mount(
			<CommonContextProviderMock>
				<ReportDetail />
			</CommonContextProviderMock>
		);
		const alert = wrapper.find('.usa-alert__heading').text();
		expect(alert).toBe('Update roster');
		wrapper.unmount();
	});
});
