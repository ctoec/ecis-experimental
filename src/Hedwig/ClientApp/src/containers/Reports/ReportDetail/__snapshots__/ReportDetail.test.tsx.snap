// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportDetail matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid-container"
  >
    <a
      class="oec-directional-link"
      href="/reports"
    >
      <svg
        class="oec-directional-link__arrow oec-directional-link__arrow--direction-left oec-directional-link__arrow--side-left"
      >
        arrowRight.svg
      </svg>
      Back to reports
    </a>
    <div
      class="usa-alert usa-alert--error grid-row margin-y-4"
      role="alert"
    >
      <div
        class="display-flex flex-fill"
      >
        <div
          class="usa-alert__body grid-col flex-fill"
        >
          <h2
            class="usa-alert__heading"
          >
            Update roster
          </h2>
          <p
            class="usa-alert__text"
          >
            There are 2 enrollments missing information required to submit this report.
          </p>
        </div>
        <div
          class="usa-alert__body flex-auto display-flex flex-align-center"
        >
          <a
            class="usa-button"
            href="/roster"
          >
            Update roster
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    class="grid-container"
  >
    <div
      class="grid-row flex-first-baseline flex-space-between"
    >
      <h1
        class="tablet:grid-col-auto"
      >
        September 2019 CDC Report 
        <span
          class="usa-tag margin-left-1 text-middle bg-gold-20v"
        >
          DRAFT
        </span>
      </h1>
      <div
        class="tablet:grid-col-auto print-btn"
      >
        <button
          class="usa-button usa-button--outline"
          type="button"
        >
          Print
        </button>
      </div>
    </div>
    <p
      class="usa-intro"
    >
      Test Organization 
      <br />
      September 01, 2019–September 28, 2019
    </p>
    <div
      class="usa-form-group margin-bottom-5"
    >
      <div
        class="margin-top-3"
      />
      <div
        class="usa-checkbox undefined"
      >
        <input
          checked=""
          class="usa-checkbox__input"
          id="accredited-accredited"
          name="accredited"
          type="checkbox"
          value="accredited"
        />
        <label
          class="usa-checkbox__label"
          for="accredited-accredited"
        >
          Accredited
        </label>
      </div>
    </div>
    <h2
      class="margin-bottom-neg-205"
    >
      Monthly utilization
    </h2>
    <table
      class="oec-table oec-table--full-width"
      id="utilization-table"
    >
      <thead>
        <tr>
          <th
            aria-sort="none"
            class="oec-table__column-header"
            role="columnheader"
            scope="col"
          >
            <span
              class="oec-table__column-title"
            />
          </th>
          <th
            aria-sort="none"
            class="oec-table__column-header"
            role="columnheader"
            scope="col"
          >
            <span
              class="oec-table__column-title"
            >
              Utilization
            </span>
          </th>
          <th
            aria-sort="none"
            class="oec-table__column-header"
            role="columnheader"
            scope="col"
          >
            <span
              class="oec-table__column-title"
            >
              Reimbursement 
              <br />
              rate (weekly)
            </span>
          </th>
          <th
            aria-sort="none"
            class="oec-table__column-header"
            role="columnheader"
            scope="col"
          >
            <span
              class="oec-table__column-title"
            >
              Total (4 weeks)
            </span>
          </th>
          <th
            aria-sort="none"
            class="oec-table__column-header"
            role="columnheader"
            scope="col"
          >
            <span
              class="oec-table__column-title"
            >
              Balance
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th>
            <strong>
              Infant/Toddler
            </strong>
             – full time
          </th>
          <td
            class=" oec-table__cell--red"
          >
            1/0 spaces
          </td>
          <td>
            $169.95
          </td>
          <td
            class=""
          >
            $0.00
          </td>
          <td
            class=" "
          >
            $0.00
          </td>
        </tr>
        <tr>
          <th>
            <strong>
              Infant/Toddler
            </strong>
             – part time
          </th>
          <td
            class=" oec-table__cell--red"
          >
            1/0 spaces
          </td>
          <td>
            $59.50
          </td>
          <td
            class=""
          >
            $0.00
          </td>
          <td
            class=" "
          >
            $0.00
          </td>
        </tr>
        <tr>
          <th>
            <strong>
              Preschool
            </strong>
             – full time
          </th>
          <td
            class=" oec-table__cell--red"
          >
            5/2 spaces
          </td>
          <td>
            $165.32
          </td>
          <td
            class=""
          >
            $1,322.56
          </td>
          <td
            class=" "
          >
            $0.00
          </td>
        </tr>
        <tr>
          <th>
            <strong>
              Preschool
            </strong>
             – part time
          </th>
          <td
            class=" oec-table__cell--red"
          >
            1/0 spaces
          </td>
          <td>
            $57.85
          </td>
          <td
            class=""
          >
            $0.00
          </td>
          <td
            class=" "
          >
            $0.00
          </td>
        </tr>
        <tr>
          <th>
            <strong>
              Total
            </strong>
          </th>
          <td
            class="oec-table__cell--strong oec-table__cell--red"
          >
            8/2 spaces
          </td>
          <td />
          <td
            class="oec-table__cell--strong"
          >
            $1,322.56
          </td>
          <td
            class="oec-table__cell--strong "
          >
            $0.00
          </td>
        </tr>
      </tbody>
    </table>
    <form
      autocomplete="off"
      class="usa-form"
      novalidate=""
    >
      <h2>
        Other Revenue
      </h2>
      <fieldset
        aria-required="true"
        class="grid-gap grid-row usa-fieldset
			
			 undefined
			"
        id="other-revenue"
      >
        <legend
          id="fieldset-legend-other-revenue"
        >
          <span
            class="usa-sr-only"
          >
            Other Revenue
          </span>
        </legend>
        <div
          class="grid-gap grid-row undefined"
        >
          <div
            class="mobile-lg:grid-col-12"
          >
            <div
              class="flex-fill usa-form-group"
            >
              <label
                class="usa-label"
                for="c4k-revenue"
              >
                <span
                  class="text-bold"
                >
                  Care 4 Kids
                </span>
                <span>
                   (0 kids receiving subsidies)
                </span>
                 
              </label>
              <input
                aria-required="true"
                class="usa-input"
                id="c4k-revenue"
                name="c4KRevenue"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="usa-form-group "
          >
            <div
              class="margin-top-3"
            />
            <div
              class="usa-checkbox undefined"
            >
              <input
                class="usa-checkbox__input"
                id="c4k-includes-retroactive-retroactiveC4KRevenue"
                name="retroactiveC4KRevenue"
                type="checkbox"
                value="retroactiveC4KRevenue"
              />
              <label
                class="usa-checkbox__label"
                for="c4k-includes-retroactive-retroactiveC4KRevenue"
              >
                Includes retroactive payment for past months
              </label>
            </div>
          </div>
          <div
            class="mobile-lg:grid-col-12"
          >
            <div
              class="flex-fill usa-form-group"
            >
              <label
                class="usa-label"
                for="family-fees-revenue"
              >
                <span
                  class="text-bold"
                >
                  Family Fees
                </span>
                 
              </label>
              <input
                aria-required="true"
                class="usa-input"
                id="family-fees-revenue"
                name="familyFeesRevenue"
                type="text"
                value="$1,000.00"
              />
            </div>
          </div>
          <div
            class="mobile-lg:grid-col-12"
          >
            <div
              class=" usa-form-group"
            >
              <label
                class="usa-label"
                for="cdc-report-comment"
              >
                <span
                  class="text-bold"
                >
                  Anything to share with the Office of Early Childhood about your report?
                </span>
                 
                <span
                  class="usa-hint"
                >
                   (optional)
                </span>
              </label>
              <textarea
                aria-required="false"
                class="usa-textarea"
                id="cdc-report-comment"
                name="comment"
              />
            </div>
          </div>
        </div>
      </fieldset>
      <button
        class="usa-button"
        disabled=""
        type="button"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
